### PR TITLE
add function to add frequently confused intent to intent report

### PR DESCRIPTION
**Pull request by amn41**
**Proposed changes**:

Adds extra info to the `intent_report.json` created by `rasa test`. Specifically, a key is added to the dict which contains up to 2 other intents with high numbers in the confusion matrix. 

Needs tests and types and general polish

output looks like this:
```
  "meta_inform_problem_bad-link": {
    "precision": 0.0,
    "recall": 0.0,
    "f1-score": 0.0,
    "support": 3,
    "confused_with": {
      "faq": 1,
      "affirm": 1
    }
  }
```
 

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/test.py [hunk 1]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
 
 
+def add_confused_intents_to_report(
+    report: Dict, target_intents: Iterable[Any], predicted_intents: Iterable[Any]
+) -> Dict:
+
+    from sklearn.metrics import confusion_matrix
+    from sklearn.utils.multiclass import unique_labels
+
+    cnf_matrix = confusion_matrix(target_intents, predicted_intents)
```
Comment (wochinge) [inline rasa/nlu/test.py]: We potentially end up calculating the confusion matrix  twice (also in line 504). Would be great if could to the confusion matrix calculating just once.

File: rasa/nlu/test.py [hunk 2]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+
+    indices = np.argsort(cnf_matrix, axis=1)
```
Comment (wochinge) [inline rasa/nlu/test.py]: Can we add comments, e.g. "Sort by confidence scores, so we can later extract the top x wrongly classified intents."

File: rasa/nlu/test.py [hunk 3]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+    labels = unique_labels(target_intents, predicted_intents)
+    n_candidates = min(3, len(labels))
+
+    for label in labels:
+        if report.get(label):
+            report[label]["confused_with"] = {}
+    for i, label in enumerate(labels):
+        for j in range(n_candidates):
+            label_idx = indices[i, -j]
+            _label = labels[label_idx]
```
Comment (wochinge) [inline rasa/nlu/test.py]: My suggestion is also really ideal, but maybe sth in this direction:
```suggestion
            label_of_confused_with = labels[label_idx]
```

File: rasa/nlu/test.py [hunk 4]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+            num_hits = int(cnf_matrix[i, label_idx])
+            if _label != label and num_hits > 0:
```
Comment (wochinge) [inline rasa/nlu/test.py]: due to the first condition you might end up with just 2 candidates even if 3 would be available.

File: rasa/nlu/test.py [hunk 5]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+    n_candidates = min(3, len(labels))
+
+    for label in labels:
+        if report.get(label):
+            report[label]["confused_with"] = {}
+    for i, label in enumerate(labels):
+        for j in range(n_candidates):
+            label_idx = indices[i, -j]
+            _label = labels[label_idx]
+            num_hits = int(cnf_matrix[i, label_idx])
```
Comment (wochinge) [inline rasa/nlu/test.py]: how about:
```suggestion
            times_the_label_was_a_false_positive= int(cnf_matrix[i, label_idx])
```

File: rasa/nlu/test.py [hunk 6]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
     }
 
 
+def add_confused_intents_to_report(
```
Comment (wochinge) [inline rasa/nlu/test.py]:
```suggestion
def _add_confused_intents_to_report(
```

File: rasa/nlu/test.py [hunk 7]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+    from sklearn.utils.multiclass import unique_labels
+
+    cnf_matrix = confusion_matrix(target_intents, predicted_intents)
+
+    indices = np.argsort(cnf_matrix, axis=1)
+    labels = unique_labels(target_intents, predicted_intents)
+    n_candidates = min(3, len(labels))
+
+    for label in labels:
+        if report.get(label):
```
Comment (wochinge) [inline rasa/nlu/test.py]: can this case actually happen? A label which is not part of the report?

File: rasa/nlu/test.py (hunk 2 above)
Comment (wochinge) [reply rasa/nlu/test.py]: same to some of the things further down

File: rasa/nlu/test.py [hunk 8]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
     }
 
 
+def add_confused_intents_to_report(
+    report: Dict, target_intents: Iterable[Any], predicted_intents: Iterable[Any]
```
Comment (wochinge) [inline rasa/nlu/test.py]:
```suggestion
    report: Dict[Text, Dict[Text, float]], target_intents: Iterable[Any], predicted_intents: Iterable[Any]
```

File: rasa/nlu/test.py [hunk 9]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+) -> Dict:
```
Comment (wochinge) [inline rasa/nlu/test.py]:
```suggestion
) -> Dict[Text, Dict[Text, Union[Dict, float]]]:
```

File: rasa/nlu/test.py [hunk 10]
```diff
@@ -398,6 +398,33 @@ def evaluate_response_selections(
...
+
+    from sklearn.metrics import confusion_matrix
```
Comment (wochinge) [inline rasa/nlu/test.py]: let's follow the convention to not directly import functions from other modules https://www.notion.so/Python-Code-Conventions-26835ffd72484dd699f28fa693916d2e

File: rasa/nlu/test.py (hunk 4 above)
Comment (amn41) [reply rasa/nlu/test.py]: yeah my assumption is that 'self' will always be in the top 3, and I'd like to report the two most commonly confused ones

File: rasa/nlu/test.py (hunk 5 above)
Comment (amn41) [reply rasa/nlu/test.py]: how about this?

File: rasa/nlu/test.py (hunk 7 above)
Comment (amn41) [reply rasa/nlu/test.py]: yes! you can predict `None` :)

File: rasa/nlu/test.py (hunk 4 above)
Comment (wochinge) [reply rasa/nlu/test.py]: Ok got it, best add this to a comment to the `min(3, candidates)`, too

File: rasa/nlu/test.py (hunk 10 above)
Comment (amn41) [reply rasa/nlu/test.py]: didn't know that, thanks!

File: rasa/nlu/test.py (hunk 7 above)
Comment (wochinge) [reply rasa/nlu/test.py]: mhm, should we add `None` to the report then?

Comment (amn41) [reply rasa/nlu/test.py]: since we report false positives, it should show up in the 'confused with', so will include that in the test, thanks!

File: rasa/nlu/test.py (hunk 4 above)
Comment (amn41) [reply rasa/nlu/test.py]: added to docstring

**Comment by amn41** [comment]
thanks @wochinge ! can you please have another quick look?

File: tests/nlu/base/test_evaluation.py [hunk 11]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
     assert result["predictions"][0] == prediction
 
 
+def test_intent_evaluation_report_large(tmpdir_factory):
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]:
```suggestion
def test_intent_evaluation_report_large(tmpdir_factory: TempdirFactory):
```

File: tests/nlu/base/test_evaluation.py [hunk 12]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+    path = tmpdir_factory.mktemp("evaluation").strpath
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]: the path comments are optional but would be great, cause we are trying to switch to `Path` now that we got rid of Python 3.5 
```suggestion
    path = tmpdir_factory.mktemp("evaluation")
```

File: tests/nlu/base/test_evaluation.py [hunk 13]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+    report_folder = os.path.join(path, "reports")
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]:
```suggestion
    report_folder = path / "reports"
```

File: tests/nlu/base/test_evaluation.py [hunk 14]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+    report_filename = os.path.join(report_folder, "intent_report.json")
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]:
```suggestion
    report_filename = report_folder / "intent_report.json"
```

File: tests/nlu/base/test_evaluation.py [hunk 15]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+
+    rasa.utils.io.create_directory(report_folder)
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]: or make `create_directory` compatible with `Path` 
```suggestion
    rasa.utils.io.create_directory(str(report_folder))
```

File: tests/nlu/base/test_evaluation.py [hunk 16]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+
+    def correct(label):
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]:
```suggestion
    def correct(label: Text) -> IntentEvaluationResult:
```

File: tests/nlu/base/test_evaluation.py [hunk 17]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+        return IntentEvaluationResult(label, label, "", 1.0)
+
+    def incorrect(label, _label):
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]:
```suggestion
    def incorrect(label: Text, _label: Text) -> IntentEvaluationResult:
```

File: tests/nlu/base/test_evaluation.py [hunk 18]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+        "support": 10,
+        "confused_with": {},
+    }
+
+    e_results = {
+        "precision": 0.0,
+        "recall": 0.0,
+        "f1-score": 0.0,
+        "support": 10,
+        "confused_with": {"C": 5, "": 5},
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]: can we have a case where we check the order (e.g. 3 times confused with A, 4 confused with B, 5 confused with C --> A shouldn't show up)

File: tests/nlu/base/test_evaluation.py [hunk 19]
```diff
@@ -320,6 +326,59 @@ def test_intent_evaluation_report(tmpdir_factory):
...
+    evaluate_intents(
+        intent_results,
+        report_folder,
+        successes=False,
+        errors=False,
+        confmat_filename=None,
+        intent_hist_filename=None,
+    )
+
+    report = json.loads(rasa.utils.io.read_file(report_filename))
```
Comment (wochinge) [inline tests/nlu/base/test_evaluation.py]: or fix `rasa.utils.io.read_file` (same as above)
```suggestion
    report = json.loads(rasa.utils.io.read_file(str(report_filename)))
```

File: rasa/nlu/test.py [hunk 20]
```diff
@@ -398,6 +398,40 @@ def evaluate_response_selections(
...
+    labels: Iterable[Text],
+) -> Dict[Text, Dict[Text, Union[Dict, float]]]:
+    """Adds a field "confused_with" to the intents in the
+    intent evaluation report. The value is a dict of
+    {"false_positive_label": false_positive_count} pairs.
+    If there are no false positives in the confusion matrix,
+    the dict will be empty. Typically we include the two most
+    commonly false positive labels, three in the rare case that
+    the diagonal element in the confusion matrix is not one of the
+    three highest values in the row"""
```
Comment (wochinge) [inline rasa/nlu/test.py]:
```suggestion
    three highest values in the row.
    """
```

File: tests/nlu/base/test_evaluation.py (hunk 18 above)
Comment (amn41) [reply tests/nlu/base/test_evaluation.py]: thanks for the suggestion! implementing this extra test caused me to discover a bug :)

File: rasa/nlu/test.py [hunk 21]
```diff
@@ -398,6 +398,41 @@ def evaluate_response_selections(
     }
 
 
+def _add_confused_intents_to_report(
+    report: Dict[Text, Dict[Text, float]],
+    cnf_matrix: np.ndarray,
+    labels: Iterable[Text],
```
Comment (wochinge) [inline rasa/nlu/test.py]: it's iterable + `Sized` (since you can apply `len(...)` to it 
```suggestion
    labels: Collection[Text],
```

File: rasa/nlu/test.py [hunk 22]
```diff
@@ -416,6 +451,8 @@ def evaluate_intents(
     Others are filtered out. Returns a dictionary of containing the
     evaluation result.
     """
+    import sklearn.metrics
```
Comment (wochinge) [inline rasa/nlu/test.py]: pycharms marks this as unused, but we don't import it anywhere else 🤔
